core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,336 @@
+                listener.get();
+                fail("Expected block but returned proper result");
+            } catch (ExecutionException ex) {
+                assertThat(ex.getCause(), equalTo(exception));
+            }
+        } else {
+            assertThat(listener.get(), equalTo(response));
+        }
+    }
+
+    @Test
+    public void testLocalOperationWithBlocks() throws ExecutionException, InterruptedException {
+        final boolean retryableBlock = randomBoolean();
+        final boolean unblockBeforeTimeout = randomBoolean();
+
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(unblockBeforeTimeout ? 60 : 1));
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        ClusterBlock block = new ClusterBlock(1, "", retryableBlock, true,
+                randomFrom(RestStatus.values()), ClusterBlockLevel.ALL);
+        ClusterState stateWithBlock = ClusterState.builder(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_MASTER))
+                .blocks(ClusterBlocks.builder().addGlobalBlock(block)).build();
+        clusterService.setState(stateWithBlock);
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+                Set<ClusterBlock> blocks = state.blocks().global();
+                return blocks.isEmpty() ? null : new ClusterBlockException(blocks);
+            }
+        }.execute(request, listener);
+
+        if (retryableBlock && unblockBeforeTimeout) {
+            Thread.sleep(TimeValue.timeValueSeconds(1).millis());
+            clusterService.setState(ClusterState.builder(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_MASTER))
+                    .blocks(ClusterBlocks.EMPTY_CLUSTER_BLOCK).build());
+
+            listener.get();
+            return;
+        }
+
+        assertListenerThrows("ClusterBlockException should be thrown", listener, ClusterBlockException.class);
+    }
+
+    @Test
+    public void testForceLocalOperation() throws ExecutionException, InterruptedException {
+        Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        clusterService.setState(mockedClusterState(
+                randomFrom(MASTER_EXISTS, MASTER_DOES_NOT_EXIST),
+                randomFrom(LOCAL_NODE_IS_NOT_MASTER, LOCAL_NODE_IS_MASTER)));
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected boolean localExecute(Request request) {
+                return true;
+            }
+        }.execute(request, listener);
+
+        listener.get();
+    }
+
+    @Test
+    public void testMasterNotAvailable() throws ExecutionException, InterruptedException {
+        Request request = new Request().masterNodeTimeout(TimeValue.timeValueSeconds(1));
```

Comment:
can we set the timeout here to 0? in general we always try to make unit tests finish as quick as possible. this one waits for 1s per run.